### ISSUE TITLE
Lib-vau-csharp fails test @vau within dc-mocks-image

### ISSUE THREAD

**Issue by CEiderEVIDENT** [gematik/lib-vau-csharp#6]
Running @VAU-Test within latest docker-image dc-mocks fails:
![csharp-vau-test](https://github.com/user-attachments/assets/f7d115a8-11dc-4b6e-aa11-0b6a071209f7)

**Comment by ghifajg** [comment]
I use also lib-vau-csharp and connecting to VAU works.

I tried the test mock and I have the same issue here:
![image](https://github.com/user-attachments/assets/4c0558c8-8dfb-4492-80d4-57db698a454a)

**Comment by CEiderEVIDENT** [comment]
Responses of system within RU are

- https://epa-as-1.dev.epa4all.de/VAU
```
Error: 400 = ¿kMessageTypeeErroriErrorCode
lErrorMessagex"verification of public keys failedkDescriptionx9Die Verifikation der Public Keys (M1) ist fehl geschlagengDetails`ÿ
``` 
- https://epa-as-2.dev.epa4all.de/VAU
```
Error: 500 = {"errorCode":"internalError","errorDetail":"M84Y 6GWQ QNU2 RYWH"}
```

**Comment by ghifajg** [comment]
@CEiderEVIDENT I found the problem: https://epa-as-1.dev.epa4all.de/VAU doesn't accept values in CBOR coded using BASE64. The question is now: Where has the fix to be done: "epa-as-1.dev.epa4all.de" or in lib-vau-csharp?

**Comment by CEiderEVIDENT** [comment]
@ghifajg 
Can you say where this BASE64 encoding is used? Is this the same error within th dc-mocks?

I think, using CBOR including BASE64-texts is allowed by the common public specification. Unfortunately this specific part is not defined in details by the gematik. obviously were only normal texts expected.

I think the specs will be changed by the gematik, so that the backends will not have to change, so lib-vau-csharp will have to be adapted.

**Comment by KillerCodeMonkey** [comment]
Maybe check this:
https://gemspec.gematik.de/docs/gemSpec/gemSpec_Krypt/gemSpec_Krypt_V2.32.0/#A_24628

It says that the inner request should be plain. So maybe this means that it should not be base64 encoded.

**Comment by ghifajg** [comment]
@CEiderEVIDENT : For M1 the "problem" starts in VauMessage1.toCBOR():

```
        public byte[] toCBOR()
        {
            return CborUtils.EncodeToCbor(this);
        }
```

the conversion to BASE64 is here in EncodeToCbor:

```
        public static byte[] EncodeToCbor(object objectToDecode)
        {
            string serializedObject = JsonConvert.SerializeObject(objectToDecode);
            return CBORObject.FromJSONString(serializedObject).EncodeToBytes();
        }
```


My quick fix for M1 is in VauMessage1:

```
       public byte[] toCBOR()
        {
            CBORObject cborVauKey = CBORObject.NewOrderedMap();     // Reihenfolge Tags wird beibehalten (prinzipiell ist die Reihenfolge egal)
            cborVauKey.Add("MessageType", "M1");
            cborVauKey.Add("ECDH_PK", CBORObject.NewMap().Add("x", this.EcdhPublicKey.X).Add("y", this.EcdhPublicKey.Y).Add("crv", "P-256"));
            cborVauKey.Add("Kyber768_PK",  this.KyberPublicKeyBytes);

            return cborVauKey.EncodeToBytes();
        }
```

The next error comes in M3 - here I'm still working on a workaround...

Regards, 
Jürgen

**Comment by ghifajg** [comment]
@CEiderEVIDENT : dc-mocks accepts BASE64. So my implementation works fine with dc-mocks but not with the real RU2 system.

**Comment by KillerCodeMonkey** [comment]
@ghifajg @CEiderEVIDENT lets see if someone will answer on this issue after two week open... (ping @Gematik-Entwicklung) or we have to wait til the consultation appointment on wednesday... or even longer

**Comment by KillerCodeMonkey** [comment]
@ghifajg but as far as i understand the c# code of the lib-vau-sharp it is just using byte arrays of cbor content what is the correct way to it when following the gemspec-Krypt.

**Comment by ghifajg** [comment]
@KillerCodeMonkey correct, but the problem is here:

```
      public static byte[] EncodeToCbor(object objectToDecode)
        {
            string serializedObject = **JsonConvert.SerializeObject(objectToDecode)**;
            return CBORObject.FromJSONString(serializedObject).EncodeToBytes();
        }
```

JsonConvert.SerializeObject(objectToDecode) converts every byte array to BASE64.

**Comment by KillerCodeMonkey** [comment]
ah i see. i did not checked what is really part of the message. but it makes sense. i see there are the kborkeys and so on that are byte arrays. makes sense then.

But then the gemspec makes no sense... everywhere, where it states "Binärwert-öffentlicher-Schlüssel-nach-keygen-Spec-Kyber768" or something similar with "Binärwert-..." in a json structure.

@ghifajg did you check the decode method, because then the answers should not be base64 encoded cbor as well :D, but if it is so it makes no sense.

**Comment by KillerCodeMonkey** [comment]
I checked the JAVA implementation this c# thingy is based on:

https://github.com/med-united/lib-vau/blob/main/src/main/java/de/gematik/vau/lib/VauClientStateMachine.java#L59
There this is called: https://github.com/med-united/lib-vau/blob/main/src/main/java/de/gematik/vau/lib/AbstractVauStateMachine.java#L62
Which ends up somewhere in the deep code of the jackson java lib:
https://github.com/FasterXML/jackson-databind/blob/2.19/src/main/java/com/fasterxml/jackson/databind/ObjectMapper.java#L4091

where the classes/objects are just transformed to json as well.

There i am a little bit lost to see if byte arrays are converted to base64, but if some of you have time maybe a little test with the jackson lib could clarify if the lib-vau / cbor utils are doing something wrong or the error would occur in java as well.

**Comment by KillerCodeMonkey** [comment]
@ghifajg seems to work with tiger-proxy in general with message 3 as well.

i added `toCbor` methods to inner message and the ordinary message 3 class.
and calling it instead of `CborUtils.EncodeToCbor`

i can try to create another pr at this repo to visualize it

**Comment by CEiderEVIDENT** [comment]
> @CEiderEVIDENT : dc-mocks accepts BASE64. So my implementation works fine with dc-mocks but not with the real RU2 system.

From my point of view anything that passes succesfully the test on dc-mocks has to work with any RU-system by definition.
DC-Mocks is the reference for any implementation, eventhough used for the KOB!

**Comment by ghifajg** [comment]
@KillerCodeMonkey Yes, "byte array" and "BASE64" work correct in the tiger-proxy. I tried also a mix of both: works also. 
BASE64 works not with https://epa-as-1.dev.epa4all.de/VAU.

https://epa-as-2.dev.epa4all.de/VAU gives another error code as @CEiderEVIDENT  mentioned above.

**Comment by ghifajg** [comment]
I also documented this in ANFPSE-2398 waiting for an answer from gematik

**Comment by KillerCodeMonkey** [comment]
@ghifajg yes i posted it already. in the gemspec nothing says, that the "data structure" we have to build is a JSON-structure. So i think this is wrong in those example implementations.

I only can imagine the tige-proxy/dc-mocks is accepting base64, because it has this json serialize/de-serialize logic to show the req, res data in the web ui, so it makes no difference for them. but this is just guessing...

**Comment by CEiderEVIDENT** [comment]
@KillerCodeMonkey
@ghifajg 

Hi guys, let's focus on a working prototyp for dc-mocks and epa-as-1.dev to have a working proof of concept.
Using this we can pass the KOB!

While doing this the gematik can find their answers, whether dc-mocks or cd-testsuite@vau or ep-as-1 or epa-as-2 is wrong!

**Comment by KillerCodeMonkey** [comment]
@CEiderEVIDENT maybe you can check the next pr: https://github.com/gematik/lib-vau-csharp/pull/9

**Comment by CEiderEVIDENT** [comment]
@KillerCodeMonkey 
https://github.com/gematik/lib-vau-csharp/pull/9 works vor

- dc-mocks
- dc-testsuite
- epa-as-1

epa-as-2 fails, but from my point of view this should be fixed in their backend

**Comment by ghifajg** [comment]
@CEiderEVIDENT please test epa-as-2 again, works now here.

**Comment by CEiderEVIDENT** [comment]
@ghifajg 
Confirmed working on epa-as-2 now

**Comment by MikeKurtze** [comment]
We have made an update to our C#- implementation. First of all we have removed all KSON references and conversions. THE CBOR format is simliar to JSON, but not identical. This would only cause confusion. Next we have made changes to the AES encryptionand decryption. The random value and the counter must be passed in as parameters, and the resulting IV is created from them. This can be retrieved using the GET method ans is also used for encryption and decryption. We have also added a test class named "EpaDeploymentTest" to test the handshake and the VAU status. We have tested this agaiunst our epa deployment environment.
